datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,137 @@
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any
+
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class ReaderThread(threading.Thread):
+    def __init__(self, file, q):
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = q
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        os.read() or queue.put(). We are closing the file that read
+        is reading from here, but the queue has to be emptied in
+        another thread in order to ensure thread-exiting.
+        """
+        self.quit = True
+        self.file.close()
+
+    def run(self):
+        logger.debug(f"ReaderThread({self.file}, {self.queue}) started")
+        while not self.quit:
+            try:
+                data = os.read(self.file.fileno(), 1024)
+            except BrokenPipeError as exc:
+                logger.debug("%s exiting (broken pipe)", self)
+                self.queue.put((self.file.fileno(), exc, time.time()))
+                return
+
+            if data == b"":
+                logger.debug(f"{self} exiting (stream end)")
+                self.queue.put((self.file.fileno(), None, time.time()))
+                return
+
+            self.queue.put((self.file.fileno(), data, time.time()))
+
+
+def run_command(cmd,
+                protocol_class,
+                stdin,
+                protocol_kwargs=None,
+                **kwargs) -> Any:
+
+    catch_stdout = protocol_class.proc_out is not None
+    catch_stderr = protocol_class.proc_err is not None
+
+    kwargs = {
+        **kwargs,
+        **dict(
+            bufsize=0,
+            stdin=stdin,
+            stdout=subprocess.PIPE if catch_stdout else None,
+            stderr=subprocess.PIPE if catch_stderr else None,
+            shell=True if isinstance(cmd, str) else False
+        )
+    }
+
+    process = subprocess.Popen(cmd, **kwargs)
+    process_stdout_fileno = process.stdout.fileno() if catch_stdout else None
+    process_stderr_fileno = process.stderr.fileno() if catch_stderr else None
+    protocol = protocol_class(**protocol_kwargs)
+
+    # We pass process as transport-argument. It does not have the same
+    # semantics as the asyncio-signature, but since it is only used in
+    # WitlessProtocol, we can change it there.
+    protocol.connection_made(process)
+
+    # Map the pipe file numbers to stdout and stderr file number, because
+    # the latter are hardcoded in the protocol code
+    # TODO: the fixed file numbers seem to be a side-effect of using
+    #  SubprocessProtocol. Some datalad-code relies on this. Shall
+    #  we replace hard coded stdout, stderr-file numbers with parameters?
```

Comment:
Given that the protocols derive from `asyncio.SubprocessProtocol`, I think sticking to the standard 1/2 make sense, so I'd vote for handling the translation in `run_command` as you currently do.

Reply:
Ok, Thanks :-)

Reply:
Hm. In my mind - the protocols don't *have* to derive from `asyncio.SubprocessProtocol` anymore with this PR.
What's important is to keep the concept, not the heritage. However - fine with me to not do that with the the first shot at it.

Reply:
> the protocols don't _have_ to derive from `asyncio.SubprocessProtocol` anymore with this PR.

I don't disagree with that, though I also don't see a benefit of decoupling.  And, on second thought, regardless of the parent class/interface, I consider mapping to the standard 1,2 fds inside `run_command` a bit nicer from the standpoint of isolating the logic/implementation details.

Reply:
We keep it as it is for now

Reply:
We go with the proposed implementation for now.